Cesium3DTilesContent/src/I3dmToGltfConverter.cpp

```diff
@@ -0,0 +1,777 @@
+    size_t i) {
+  copyToBuffer(position, rotation, scale, &bufferData[i * totalStride]);
+}
+
+void instantiateGltfInstances(
+    GltfConverterResult& result,
+    const DecodedInstances& decodedInstances) {
+  std::set<CesiumGltf::Node*> meshNodes;
+  int32_t instanceBufferId = createBufferInGltf(*result.model);
+  auto& instanceBuffer =
+      result.model->buffers[static_cast<uint32_t>(instanceBufferId)];
+  int32_t instanceBufferViewId = createBufferViewInGltf(
+      *result.model,
+      instanceBufferId,
+      0,
+      static_cast<int64_t>(totalStride));
+  auto& instanceBufferView =
+      result.model->bufferViews[static_cast<uint32_t>(instanceBufferViewId)];
+  const auto numInstances =
+      static_cast<uint32_t>(decodedInstances.positions.size());
+  auto upToZ = CesiumGltfContent::GltfUtilities::applyGltfUpAxisTransform(
+      *result.model,
+      glm::dmat4x4(1.0));
+  result.model->forEachPrimitiveInScene(
+      -1,
+      [&](Model& gltf,
+          Node& node,
+          Mesh&,
+          MeshPrimitive&,
+          const glm::dmat4& transform) {
+        auto [nodeItr, notSeen] = meshNodes.insert(&node);
+        if (!notSeen) {
+          return;
+        }
+        std::vector<glm::dmat4> existingInstances{glm::dmat4(1.0)};
+        auto& gpuExt = node.addExtension<ExtensionExtMeshGpuInstancing>();
+        if (!gpuExt.attributes.empty()) {
+          // The model already has instances! We will need to create the outer
+          // product of these instances and those coming from i3dm.
+          existingInstances = getMeshGpuInstancingTransforms(
+              *result.model,
+              gpuExt,
+              result.errors);
+          if (numInstances * existingInstances.size() >
+              std::numeric_limits<uint32_t>::max()) {
+            result.errors.emplaceError(fmt::format(
+                "Too many instances: {} from i3dm and {} from glb",
+                numInstances,
+                existingInstances.size()));
+            return;
+          }
+        }
+        if (result.errors.hasErrors()) {
+          return;
+        }
+        const uint32_t numNewInstances =
+            static_cast<uint32_t>(numInstances * existingInstances.size());
+        const size_t instanceDataSize = totalStride * numNewInstances;
+        auto dataBaseOffset =
+            static_cast<uint32_t>(instanceBuffer.cesium.data.size());
+        instanceBuffer.cesium.data.resize(dataBaseOffset + instanceDataSize);
+        // Transform instance transform into local glTF coordinate system.
+        const auto toTile = upToZ * transform;
+        const auto toTileInv = inverse(toTile);
+        size_t destInstanceIndx = 0;
+        for (unsigned i = 0; i < numInstances; ++i) {
+          auto instMat = composeInstanceTransform(i, decodedInstances);
+          instMat = toTileInv * instMat * toTile;
+          for (const auto& innerMat : existingInstances) {
+            auto finalMat = instMat * innerMat;
```

Comment:
I think some name tweaks would help with readability:

```suggestion
        const glm::dmat4 toTile = upToZ * transform;
        const glm::dmat4 toTileInv = inverse(toTile);
        size_t destInstanceIndx = 0;
        for (unsigned i = 0; i < numInstances; ++i) {
          glm::dmat4 i3dmTransform = composeInstanceTransform(i, decodedInstances);
          glm::dmat4 i3dmToGltf = toTileInv * i3dmTransform* toTile;
          for (const auto& gltfTransform : existingInstances) {
            glm::dmat4 finalTransform = i3dmToGltf * gltfTransform;
```

Not sure if `i3dmToGltf` is the exact correct name, but something similar would help

Reply:
I've changed the names and reworked this a bit.